yjit/src/asm/x86_64/mod.rs

```diff
@@ -617,451 +645,546 @@ impl CodeBlock
         cb->num_refs = 0;
     }
     */
-}
 
-/*
-// Check if an operand needs a REX byte to be encoded
-static bool rex_needed(x86opnd_t opnd)
-{
-    if (opnd.type == OPND_NONE || opnd.type == OPND_IMM)
-    {
-        return false;
-    }
+    fn mark_position_writeable(&mut self, write_pos: usize) {
+        let page_size = page_size();
+        let aligned_position = (self.write_pos / page_size) * page_size;
 
-    if (opnd.type == OPND_REG)
-    {
-        return (
-            opnd.as.reg.reg_no > 7 ||
-            (opnd.num_bits == 8 && opnd.as.reg.reg_no >= 4 && opnd.as.reg.reg_no <= 7)
-        );
+        if self.current_aligned_write_pos != aligned_position {
+            self.current_aligned_write_pos = aligned_position;
+            self.mem_block.mark_read_write(aligned_position, page_size).unwrap();
+        }
     }
 
-    if (opnd.type == OPND_MEM)
-    {
-        return (opnd.as.mem.base_reg_no > 7) || (opnd.as.mem.has_idx && opnd.as.mem.idx_reg_no > 7);
+    fn mark_all_executable(&mut self) {
+        self.current_aligned_write_pos = ALIGNED_WRITE_POSITION_NONE;
+        self.mem_block.mark_executable(0, self.mem_size).unwrap();
     }
 
-    rb_bug("unreachable");
-}
-
-// Check if an SIB byte is needed to encode this operand
-static bool sib_needed(x86opnd_t opnd)
-{
-    if (opnd.type != OPND_MEM)
-        return false;
-
-    return (
-        opnd.as.mem.has_idx ||
-        opnd.as.mem.base_reg_no == RSP.as.reg.reg_no ||
-        opnd.as.mem.base_reg_no == R12.as.reg.reg_no
-    );
-}
-
-// Compute the size of the displacement field needed for a memory operand
-static uint32_t disp_size(x86opnd_t opnd)
-{
-    assert (opnd.type == OPND_MEM);
+    /// Write the REX byte
+    fn write_rex(&mut self, w_flag: bool, reg_no: u8, idx_reg_no: u8, rm_reg_no: u8) {
```

Comment:
IIRC we had discussed this and it would make the initial porting easier to have these instruction generation functions not be methods on an Assembler trait, but remain individual functions. Unless there's a specific advantage we're going after by doing that.